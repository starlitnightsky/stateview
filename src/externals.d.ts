declare module '*.less' {
  const resource: { [key: string]: string };
  export = resource;
}

declare interface LayerProps {
  state: string;
  component?: any;
  children?: any;
  data?: any;
}
