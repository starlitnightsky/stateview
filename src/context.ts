import { createContext, useContext } from "react"

export type GlobalStateContent = {
  stateContent: Object,
  stateview: any,
  setStateview: (c: any) => void
}

export const StateContext = createContext<GlobalStateContent>({
  stateContent: {}, // set a default value
  stateview: null,
  setStateview: () => { },
})

export const useStateContext = () => useContext(StateContext)
