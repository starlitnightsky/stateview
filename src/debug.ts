import IDebug from 'debug';

export const Debug = (content: any) => {
  return IDebug(content);
};
