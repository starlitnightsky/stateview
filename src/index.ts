
import type { IState } from './type';

declare global {
    interface Window {
        stateview: IState
    }
}

export * from './type' 
export * from './stateview'
export * from './layer'
export * from './context'
export * from './state'
export * from './debug' 