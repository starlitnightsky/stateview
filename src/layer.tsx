import React, { Fragment } from 'react';

declare interface LayerProps {
    state: string;
    component?: any;
    children?: any;
    data?: any;
}

export function Layer(props: LayerProps) {
    return <Fragment>{props.component}</Fragment>;
}
