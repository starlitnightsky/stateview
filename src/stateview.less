
.normal {
}

.stateview {
  // background: rgb(121, 242, 157);
  // border: 1px dashed lightblue;
}

.example {
  // background: rgb(121, 242, 157);
  // border: 1px dashed red;
}

