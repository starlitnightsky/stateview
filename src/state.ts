// import { createContext, useRef } from "react"
import type { ICconfig, IState } from './type';

import { Debug } from './debug';

const debug = Debug("src/state.ts")
// 获取对象
//  1. React.Children.forEach(statview.props.children）
//       注册状态和Layer，statview-ref
//  2. 实现statview的行为
//       stateview.setCurrent(this | 'somepath')
//       stateview.keyframes(from , to)
//       ...
//  3. 通过useStateContext，可以在Layer中使用
// 
// 示例如下:
// 
// function Welcome(props: any) {
//     const { stateContent, stateview } = useStateContext()
//     console.dir("Welcome")
//     console.dir(stateContent)
// 
//     // stateview.setCurrent(this | 'somepath')
//     // stateview.keyframes(from , to)
// 
//     return <h1>Hello, { props.name } < /h1>;
// }
//  



export const SState = (config: ICconfig) => {
    // 处理配置

    const states = Object.keys(config.GlobalStateMapping)

    // 对外行为
    return Object.assign(config, {
        _vshow: function (path: string, c: any) {
            let state = this._getStateBy(path);
            var setVisibaleComponent = state.show
            setVisibaleComponent(c)
        },
        _setCurrent(path: string) {
            config.GlobalStateMapping.currentState = path
        },
        _getStateBy(path: string) {
            let state = config.GlobalStateMapping[path] as IState;
            return state
        },
        /**
         * 根据path，显示path对应的Layer
         */
        show: function (path: string) {
            let state = this._getStateBy(path);
            // 如果状态一样，则无需刷新
            if (path === state.currentState) {
                console.warn("状态一样，无需刷新")
                return
            }
            // 确保状态在states中，完成show切换
            if (states.indexOf(path) !== -1) {
                // TODO：根据path，获得该path归属的stateview
                // 状态嵌套
                debug(path)

                // 根据状态获取视图
                let visibaleComponent = state.component

                // 设置stateview渲染view
                this._vshow(path, visibaleComponent)

                // 设置当前的状态
                this._setCurrent(path)
            } else {
                console.warn("setCurrent()参数中的" + path + " 状态不在stateview中[" + states.join('/') + "]")
            }
        }
    })
}


