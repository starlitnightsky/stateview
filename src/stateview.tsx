import React, { useEffect, useState, useRef } from 'react';
import { Debug } from './debug';
import { getQueryStringByName } from './utils';
import { GlobalStateContent, StateContext } from './context';
import { SState } from './state';

import type { ICconfig } from './type';
import styles from './stateview.less';

const debug = Debug("stateview.jsx")

var count: number = 0;
var GlobalStateMapping: any = {}

export const Stateview = React.forwardRef((props: any, ref: any) => {
    const [visibaleComponent, setVisibaleComponent] = useState(0);
    // const states = getStates(props.children); 
    let states: any = []
    // states.for()
    React.Children.forEach(props.children, child => {
        // const childType = { ...child.type }
        debug(child)
        // console.log(child.name)
        debug(child.props.router)
        states.push(child.props.router)

        let c = child.props.component

        if (!child.props.component) {
            c = child.props.children
        }

        GlobalStateMapping[child.props.router] = {
            show: setVisibaleComponent,
            child: child,
            component: c
        }
    })

    const stateviewRef = useRef(null);

    const sState = SState({
        ref: stateviewRef,
        GlobalStateMapping: GlobalStateMapping
    } as ICconfig)

    if (count === 0) {
        GlobalStateMapping.currentState = props.default
        window.stateview = sState
    }

    debug(count++)
    debug(sState)

    const ctx = {
        stateContent: props.children,
        stateview: sState
    } as GlobalStateContent;

    useEffect(() => {
        // show default
        const d = GlobalStateMapping[props.default]
        d.show(d.component)
    }, [])

    let i = getQueryStringByName('s')
    debug(props.children[i])

    let isBlock = props.block ? true : false

    return (
        <StateContext.Provider value={ctx} >
            {isBlock
                ? <span ref={ref} className={styles.example} style={{ height: props.height }}>
                    {visibaleComponent}
                </span>
                : <div ref={ref} className={styles.example} style={{ height: props.height }}>
                    {visibaleComponent}
                </div>
            }
        </StateContext.Provider>
    );
})
