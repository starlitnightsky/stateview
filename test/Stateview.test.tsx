import { shallow } from 'enzyme';
import React from 'react';
// import { act } from 'react-dom/test-utils';
import { Stateview, Layer } from '../src';
// import type { ExtendablePromise } from '../src/types';

const Example1 = () => {

    function unlogin() {
        window.stateview.show('unlogin')
    }

    function logined() {
        window.stateview.show('logined')
    }

    return (
        <Stateview default='unlogin'>
            <Layer router='logined'>
                <h1>Logined, <button onClick={unlogin}>go to UnLogin</button></h1>
            </Layer>
            <Layer router='unlogin'>
                <h1 >UnLogin, <button onClick={logined}>go to Logined</button></h1>;
            </Layer>
        </Stateview>
    );
}
// const waitForComponentToPaint = async (wrapped: ReactWrapper) => {
//   await act(async () => {
//     await new Promise((resolve) => setTimeout(resolve, 0));
//     wrapped.update();
//   });
// };

describe('<Stateview /> component', () => {
    describe('Truthy cases', () => {
        describe('With NODE_ENV === test', () => {
            test('GIVEN <Then /> THEN renders children', () => {
                const wrapped = shallow(
                    <Example1 />
                );

                expect(wrapped).toMatchSnapshot();
                expect(wrapped.containsMatchingElement(<h1>UnLogin, <button>go to Logined</button></h1>)).toBe(true);
            });
        });
    });
});
