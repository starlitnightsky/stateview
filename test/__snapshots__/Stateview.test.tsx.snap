// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stateview /> component Truthy cases With NODE_ENV === test GIVEN <Then /> THEN renders children 1`] = `
<ForwardRef
  default="unlogin"
>
  <Layer
    router="logined"
  >
    <h1>
      Logined, 
      <button
        onClick={[Function]}
      >
        go to UnLogin
      </button>
    </h1>
  </Layer>
  <Layer
    router="unlogin"
  >
    <h1>
      UnLogin, 
      <button
        onClick={[Function]}
      >
        go to Logined
      </button>
    </h1>
    ;
  </Layer>
</ForwardRef>
`;
